dali/operators/signal/decibel/to_decibels.cc

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/operators/signal/decibel/to_decibels.h"
+#include "dali/core/static_switch.h"
+#include "dali/kernels/signal/decibel/to_decibels_cpu.h"
+#include "dali/pipeline/data/views.h"
+
+#define TO_DB_SUPPORTED_TYPES (float)
+#define TO_DB_SUPPORTED_NDIMS (1, 2, 3, 4)
+
+static constexpr int kNumInputs = 1;
+static constexpr int kNumOutputs = 1;
+
+namespace dali {
+
+DALI_SCHEMA(ToDecibels)
+    .DocStr(R"code(Converts a magnitude (real, positive) to the decibel scale, according to the
+formula:
+```
+    out[i] = multiplier * log10( max(min_ratio, input[i] / reference) )
+```)code")
+    .NumInput(kNumInputs)
+    .NumOutput(kNumOutputs)
+    .AddOptionalArg("multiplier",
+      R"code(Factor by which we multiply the logarithm (typically 10.0 or 20.0 depending if we
+are dealing with a squared magnitude or not).)code",
+      10.0f)
+    .AddOptionalArg("reference",
+      R"code(Reference magnitude. If not provided, the maximum of the input will be used as
+reference. Note: The maximum of the input will be calculated on a per-sample basis.)code",
+      1.0f)
+    .AddOptionalArg("min_ratio",
+      R"code(Minimum or cut-off ratio. Any ratio below this value will saturate. Example:
+A value `min_ratio=1e-8` corresponds to a dB cut-off of -80 dBs (assuming `multiplier=10.0`).)code",
+      1e-8f);
+
+template <>
+bool ToDecibels<CPUBackend>::SetupImpl(std::vector<OutputDesc> &output_desc,
+                                       const workspace_t<CPUBackend> &ws) {
+  output_desc.resize(kNumOutputs);
+  const auto &input = ws.InputRef<CPUBackend>(0);
+  auto &output = ws.OutputRef<CPUBackend>(0);
+  kernels::KernelContext ctx;
+  auto in_shape = input.shape();
+  int nsamples = input.size();
+  auto nthreads = ws.GetThreadPool().size();
+
+  TYPE_SWITCH(input.type().id(), type2id, T, TO_DB_SUPPORTED_TYPES, (
+    VALUE_SWITCH(in_shape.sample_dim(), Dims, TO_DB_SUPPORTED_NDIMS, (
+      using ToDbKernel = kernels::signal::ToDecibelsCpu<T, Dims>;
+      kmgr_.Initialize<ToDbKernel>();
```

Comment:
It seems as a common pattern for pass through operators - where input and output matches. Maybe we can extract it to some common place.

Reply:
As we agreed, it can't be easily generalized to the point of being reusable